src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,759 @@
+
+		if (querySent == 0)
+		{
+			ReportConnectionError(connection, ERROR);
+		}
+	}
+
+	/* Process the result */
+	foreach_ptr(connection, nodeConnectionList)
+	{
+		bool raiseInterrupts = true;
+		PGresult *result = GetRemoteCommandResult(connection, raiseInterrupts);
+
+		if (!IsResponseOK(result))
+		{
+			ereport(ERROR,
+					(errmsg("Internal error, connection failure")));
+		}
+
+		PQclear(result);
+		ForgetResults(connection);
+	}
+}
+
+
+/*
+ * During prepare, once all the nodes acknowledge commit, persist the current
+ * transaction id along with the clock value in the catalog.
+ */
+void
+PrepareAndSetTransactionClock(List *transactionNodeList)
+{
+	if (!EnableClusterClock)
+	{
+		/* citus.enable_cluster_clock is false */
+		return;
+	}
+
+	/*
+	 * Call get_current_transaction_id UDF to get the current
+	 * distributed transaction id.
+	 */
+	Oid transactionFuncOid = FunctionOid("pg_catalog", "get_current_transaction_id", 0);
+	Datum transactionIdHeapDatum = OidFunctionCall0(transactionFuncOid);
+
+	FmgrInfo *outputFunction = (FmgrInfo *) palloc0(sizeof(FmgrInfo));
+	Oid outputFunctionId = FunctionOid("pg_catalog", "record_out", 1);
+	fmgr_info(outputFunctionId, outputFunction);
+	char *transactionId = OutputFunctionCall(outputFunction, transactionIdHeapDatum);
+
+	/* Pick the highest logical clock value among all transaction-nodes */
+	uint64 transactionClockValue = GetHighestClockInTransaction(transactionNodeList);
+
+	/* Persist the transactionId along with the logical commit-clock timestamp */
+	LogTransactionCommitClock(transactionId, transactionClockValue);
+
+	/* Adjust all the nodes with the new clock value */
+	AdjustClocksToTransactionHighest(transactionNodeList, transactionClockValue);
+}
+
+
+/*
+ * LogTransactionCommitClock registers the committed transaction along
+ * with the commit clock.
+ */
+static void
+LogTransactionCommitClock(char *transactionId, uint64 transactionClockValue)
+{
+	Datum values[Natts_pg_dist_commit_transaction];
+	bool isNulls[Natts_pg_dist_commit_transaction];
+	uint64 clockLogical = GET_LOGICAL(transactionClockValue);
+	uint32 clockCounter = GET_COUNTER(transactionClockValue);
+
+	ereport(DEBUG1, (errmsg("Persiting transaction %s with "
+							"clock logical (%lu) and counter(%u)",
+							transactionId, clockLogical, clockCounter)));
+
+	/* form new transaction tuple */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	uint64 transactionTimestamp = GetEpochTimeMs();
+
+	values[Anum_pg_dist_commit_transaction_transactionid - 1] =
+		CStringGetTextDatum(transactionId);
+	values[Anum_pg_dist_commit_transaction_clock_logical - 1] =
+		Int64GetDatum(clockLogical);
+	values[Anum_pg_dist_commit_transaction_clock_counter - 1] =
+		Int64GetDatum(clockCounter);
+	values[Anum_pg_dist_commit_transaction_timestamp - 1] =
+		Int64GetDatum(transactionTimestamp);
+
+	/* open transaction relation and insert new tuple */
```

Comment:
just to be more clear;

/* open pg_dist_commit_transaction and insert the new tuple */